Adding alternative downsampling examples - Using Quix (#5451)
influxdata/docs-v2: content/influxdb/clustered/process-data/downsample/quix.md

---
title: Downsample data stored in InfluxDB using Quix Streams
description: >
  Use [Quix Streams](https://github.com/quixio/quix-streams) to query time series
  data stored in InfluxDB and written to Kafka at regular intervals, continuously
  downsample it, and then write the downsampled data back to InfluxDB.
menu:
  influxdb_clustered:
    name: Use Quix
    parent: Downsample data
    identifier: downsample-quix
weight: 202
related:

>   - /influxdb/cloud-serverless/query-data/sql/aggregate-select/, Aggregate or apply selector functions to data (SQL)

Reviewer: ```suggestion
  - /influxdb/clustered/query-data/sql/aggregate-select/, Aggregate or apply selector functions to data (SQL)
```

## Pipeline architecture

The following diagram illustrates how data is passed between processes as it is downsampled:

{{< html-diagram/quix-downsample-pipeline >}}

{{% note %}}
It is usually more efficient to write raw data directly to Kafka rather than
writing raw data to InfluxDB first (essentially starting the Quix Streams
pipeline with the "raw-data" topic). However, this guide assumes that you
already have raw data in InfluxDB that you want to downsample.
{{% /note %}}

---

1.  [Set up prerequisites](#set-up-prerequisites)
2.  [Install dependencies](#install-dependencies)
3.  [Prepare InfluxDB buckets](#prepare-influxdb-buckets)
4.  [Create the downsampling logic](#create-the-downsampling-logic)
5.  [Create the producer and consumer clients](#create-the-producer-and-consumer-clients)
    1. [Create the producer](#create-the-producer)
    2. [Create the consumer](#create-the-consumer)
6.  [Get the full downsampling code files](#get-the-full-downsampling-code-files)

## Set up prerequisites

The process described in this guide requires the following:

> - An InfluxDB Cloud Serverless account with data ready for downsampling.

Reviewer: ```suggestion
- An InfluxDB cluster with data ready for downsampling.
```

- A [Quix Cloud](https://portal.platform.quix.io/self-sign-up/) account or a
  local Apache Kafka or Red Panda installation.
- Familiarity with basic Python and Docker concepts.

## Install dependencies

Use `pip` to install the following dependencies:

- `influxdb_client_3`
- `quixstreams<2.5`
- `pandas`

```sh
pip install influxdb3-python pandas quixstreams<2.5
```

## Prepare InfluxDB buckets

> The downsampling process involves two InfluxDB buckets.
> Each bucket has a [retention period](/influxdb/cloud-serverless/reference/glossary/#retention-period)
> that specifies how long data persists before it expires and is deleted.
> By using two buckets, you can store unmodified, high-resolution data in a bucket
> with a shorter retention period and then downsampled, low-resolution data in a
> bucket with a longer retention period.
> 
> Ensure you have a bucket for each of the following:
> 
> - One to query unmodified data from
> - The other to write downsampled data to
> 
> For information about creating buckets, see
> [Create a bucket](/influxdb/cloud-serverless/admin/buckets/create-bucket/).

Reviewer: ```suggestion
The downsampling process involves two InfluxDB databases.
Each database has a [retention period](/influxdb/clustered/reference/glossary/#retention-period)
that specifies how long data persists before it expires and is deleted.
By using two databases, you can store unmodified, high-resolution data in a database
with a shorter retention period and then downsampled, low-resolution data in a
database with a longer retention period.

Ensure you have a database for each of the following:

- One to query unmodified data from
- The other to write downsampled data to

For information about creating databases, see
[Create a bucket](/influxdb/clustered/admin/databases/create/).
```

    sdf = (
        sdf.apply(lambda value: value[target_field])  # Extract temperature values
        .tumbling_window(timedelta(minutes=1))   # 1-minute tumbling windows
        .mean()                                  # Calculate average temperature
        .final()                                 # Emit results at window completion
    )

    sdf = sdf.apply(
        lambda value: {
            'time': value['end'],                  # End of the window
            'temperature_avg': value['value'],     # Average temperature
        }
    )

    sdf.to_topic(output_topic) # Output results to the 'downsampled-data' topic
    # ...
    ```

The results are streamed to the Kafka topic, `downsampled-data`.

{{% note %}}
Note: "sdf" stands for "Streaming Dataframe".
{{% /note %}}

You can find the full code for this process in the
[Quix GitHub repository](https://github.com/quixio/template-influxdbv3-downsampling/blob/dev/Downsampler/main.py).

## Create the producer and consumer clients

Use the `influxdb_client_3` and `quixstreams` modules to  instantiate two clients that interact with InfluxDB and Apache Kafka:

> - A **producer** client configured to read from your InfluxDB bucket with _unmodified_ data and _produce_ that data to Kafka.
> - A **consumer** client configured to _consume_ data from Kafka and write the _downsampled_ data to the corresponding InfluxDB bucket.

Reviewer: ```suggestion
- A **producer** client configured to read from your InfluxDB database with _unmodified_ data and _produce_ that data to Kafka.
- A **consumer** client configured to _consume_ data from Kafka and write the _downsampled_ data to the corresponding InfluxDB database.
```

### Create the producer client

Provide the following credentials for the producer:

> - **host**: [{{< product-name >}} region URL](/influxdb/cloud-serverless/reference/regions)
>   _(without the protocol)_
> - **org**: InfluxDB organization name
> - **token**: InfluxDB API token with read and write permissions on the buckets you
>   want to query and write to.
> - **database**: InfluxDB bucket name

Reviewer: ```suggestion
- **host**: {{< product-name omit=" Clustered">}} cluster URL
  _(without the protocol)_
- **org**: An arbitrary string. {{< product-name >}} ignores the organization.
- **token**: InfluxDB database token with read and write permissions on the databases you
  want to query and write to.
- **database**: InfluxDB database name
```

The producer queries for fresh data from InfluxDB at specific intervals. It's configured to look for a specific measurement defined in a variable. It writes the raw data to a Kafka topic called 'raw-data'

> {{% code-placeholders "(API|(RAW|DOWNSAMPLED)_BUCKET|ORG)_(NAME|TOKEN)" %}}

Reviewer: ```suggestion
{{% code-placeholders "(RAW|DOWNSAMPLED)_DATABASE_(NAME|TOKEN)" %}}
```

```py
from influxdb_client_3 import InfluxDBClient3
from quixstreams import Application
from quixstreams.models.serializers.quix import JSONSerializer, SerializationContext
import pandas

> # Instantiate an InfluxDBClient3 client configured for your unmodified bucket

Reviewer: ```suggestion
# Instantiate an InfluxDBClient3 client configured for your unmodified database
```

influxdb_raw = InfluxDBClient3(
    host='{{< influxdb/host >}}',

>     token='API_TOKEN',
>     database='RAW_BUCKET_NAME'

Reviewer: ```suggestion
    token='DATABASE_TOKEN',
    database='RAW_DATABASE_NAME'
```

    """
    #... remaining code trunctated for brevity ...

            for index, obj in enumerate(records):
                print(obj) # Obj contains each row in the table includimng temperature
                # Generate a unique message_key for each row
                message_key = obj['machineId']
                logger.info(f'Produced message with key:{message_key}, value:{obj}')

                serialized = topic.serialize(
                    key=message_key, value=obj, headers={'uuid': str(uuid.uuid4())}
                    )

                # publish each row returned in the query to the topic 'raw-data'
                producer.produce(
                    topic=topic.name,
                    headers=serialized.headers,
                    key=serialized.key,
                    value=serialized.value,
                )

```
{{% /code-placeholders %}}

You can find the full code for this process in the
[Quix GitHub repository](https://github.com/quixio/template-influxdbv3-downsampling/blob/dev/InfluxDB%20V3%20Data%20Source/main.py).

### Create the consumer

As before, provide the following credentials for the consumer:

> - **host**: [{{< product-name >}} region URL](/influxdb/cloud-serverless/reference/regions)
>   _(without the protocol)_
> - **org**: InfluxDB organization name
> - **token**: InfluxDB API token with read and write permissions on the buckets you
>   want to query and write to.
> - **database**: InfluxDB bucket name

Reviewer: ```suggestion
- **host**: {{< product-name omit=" Clustered">}} cluster URL
  _(without the protocol)_
- **org**: An arbitrary string. {{< product-name >}} ignores the organization.
- **token**: InfluxDB database token with read and write permissions on the databases you
  want to query and write to.
- **database**: InfluxDB database name
```

This process reads messages from the Kafka topic `downsampled-data` and writes each message as a point dictionary back to InfluxDB.

> {{% code-placeholders "(API|(RAW|DOWNSAMPLED)_BUCKET|ORG)_(NAME|TOKEN)" %}}

Reviewer: ```suggestion
{{% code-placeholders "(RAW|DOWNSAMPLED)_DATABASE_(NAME|TOKEN)" %}}
```

```py
# Instantiate an InfluxDBClient3 client configured for your downsampled database.
# When writing, the org= argument is required by the client (but ignored by InfluxDB).
influxdb_downsampled = InfluxDBClient3(
    host='{{< influxdb/host >}}',

>     token='API_TOKEN',
>     database='DOWNSAMPLED_BUCKET_NAME',

Reviewer: ```suggestion
    token='DATABASE_TOKEN',
    database='DOWNSAMPLED_DATABASE_NAME',
```
